text/0000-reseeding-rand.md

```diff
@@ -0,0 +1,682 @@
+- `rng.gen_ascii_chars()`, this seems relatively specialised, and the
+  full range can be simulated with `rng.gen_iter('0'..'{')` (NB. `{`
+  is the char after `z`, necessary due to exclusivity of the RHS)
+  which unfortunately picks up some punctuation in the middle, but
+  most use cases will be perfectly well served by just
+  `rng.gen_iter('a'..'{')`. We could also provide extra
+  implementations, such as `struct PrintableAscii; impl
+  Random<PrintableAscii> for char`, allowing
+  `rng.gen_iter(PrintableAscii)`.
+- `rng.choose(slice)`, this is
+  `slice.get(rng.gen(0..slice.len()))`. Its signature of `fn
+  choose<T>(&self, &[T]) -> Option<&T>` seems insufficiently general
+  to be worth including, given how simple it is to reimplement.
+
+#### Alternatives
+
+This trait doesn't need to exist, everything could be free functions
+in the `rand` module.
+
+There could be a `fn by_ref<'a>(&'a mut self) -> ByRef<'a, Self>`
+function, like occurs for iterators and in IO, to provide a real type
+backing up `impl<'a, T: ?Sized + Rng> Rng for &'a mut T {}`.
+
+## Free functions
+
+[free functions]: #free-functions
+
+Top level convenience functions are provided:
+
+```rust
+fn thread_rng() -> ThreadRng {
+    // fetch the cached RNG from TLS, or create it if it doesn't exist
+}
+
+fn random<T: Random<D>, D>(data: D) -> T {
+    thread_rng().gen(data)
+}
+```
+
+### Alternatives
+
+There is valid concern that the choice of RNGs described below leaves
+users looking at jargony names for RNGs, which no particularly obvious
+default choice; possibly leading to poor selection. Functions like:
+
+```rust
+fn secure_rng() -> OsRng { ... }
+fn balanced_rng() -> ChaChaRng { ... }
+fn weak_rng() -> XorShiftRng { ... }
+```
+
+could be provided to give some guidance. As a counterpoint, the
+`collections` library does not provide `ordered_map() -> BTreeMap`,
+`unordered_map() -> HashMap`.
+
+## Error handling
+
+[error handling]: #error-handling
+
+As it stands, there is no error handling at all, which is
+unfortunate. Most PRNGs do not have any error conditions (some define
+reaching the end of the sequence and repeating as an error), but the
+interface with the operating system may.
+
+Reconciling this is similar to IO error handling, where some readers
+and writers are guaranteed to not return errors (e.g. reading from
+memory). This consideration is postponed until the design in
+[RFC 576](https://github.com/rust-lang/rfcs/pull/576) is definite, but
+it would likely be along the lines of making each function return
+`Result<..., <Self as Reader>::Error>`. instead of just `...`. This
+may have unfortunate ergonomic impact, and may be more complicated if
+building on `Iterator<u32>`.
+
+## Provided RNGs
+
+[provided rngs]: #provided-rngs
+
+The module will provide 3 basic algorithms to try to provide more clarity:
+
+- `OsRng`, a "true" random number generator calling into the operating system,
```

Comment:
OpenBSD implements the true best practices here, and that means using `arc4random` (a userspace chacha20 implementation, seeded and occasionally reseeded from `getentropy`) rather than directly using the slow and non-scalable (global locking) OS interface.